### PR TITLE
ChangeFeedProcessor: Fixes NullReferenceException on partition split when lease has no FeedRange 

### LINKED ISSUE #6076: Change Feed Processor (in-memory lease container): NullReferenceException on partition split when leases are rehydrated without a `FeedRange`
## Summary

When the Change Feed Processor is configured with an **in-memory lease container** and is started from **previously saved/restored lease state**, a subsequent **physical partition split** throws a `NullReferenceException` (or `InvalidCastException`) inside the split handler **before** child leases are created and **before** the parent lease is deleted. The error is caught and re-notified in a loop, so:

- the parent (now-gone) partition lease is never removed,
- no child-partition leases are created,
- the processor stops making progress on the affected range,
- recovery only happens after a **full service restart** (which re-bootstraps leases from scratch).

With a **persistent (Cosmos-backed) lease container** the same scenario works, because reloaded leases carry a populated `FeedRange`.

## Impact

- **Severity:** High for users of the in-memory lease container that persist/restore lease state across restarts (the documented usage pattern for in-memory leases).
- **Blast radius:** The change feed silently stops advancing for the split partition until the process is restarted; data processing is delayed/halted for that range.
- **Regression context:** Reported by a customer after migrating from the V2 SDK (which handled in-memory lease split/cleanup) to V3, where in-memory lease support was re-added in stages.

## Environment

- SDK: `Microsoft.Azure.Cosmos` v3 (repro against current `master`/msdata branch)
- Feature: Change Feed Processor with **in-memory lease container** + restore from saved lease state
- Connectivity mode: N/A (lease-management layer)

## Root Cause (code-level)

The split handler unconditionally casts `lease.FeedRange` to `FeedRangeEpk` and dereferences it:

`Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs` — `HandlePartitionGoneAsync`:

```csharp
IReadOnlyList<PartitionKeyRange> overlappingRanges = await this.partitionKeyRangeCache.TryGetOverlappingRangesAsync(
    this.containerRid,
    ((FeedRangeEpk)lease.FeedRange).Range,   // NRE / InvalidCast when lease.FeedRange is null
    NoOpTrace.Singleton,
    forceRefresh: true);
```

`DocumentServiceLeaseCore.FeedRange` is serialized with `NullValueHandling = NullValueHandling.Ignore`:

`Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/LeaseManagement/DocumentServiceLeaseCore.cs`:

```csharp
[JsonProperty("FeedRange", NullValueHandling = NullValueHandling.Ignore)]
public override FeedRangeInternal FeedRange { get; set; }
```

So a **partition-key-range-id–style lease** (token = PKRange id, no `FeedRange`) that is **rehydrated from saved in-memory state** (`DocumentServiceLeaseStoreManagerInMemory.DeserializeLeaseState` → `InMemoryLeaseJsonFormat.Deserialize`) comes back with `FeedRange == null`. On the next split, the cast/dereference above throws.

The exception is caught by `PartitionControllerCore.HandlePartitionGoneAsync` and turned into an error notification. Because the throw happens **before** child-lease creation and **before** `leaseManager.DeleteAsync(parentLease)`, neither cleanup nor child creation ever runs, and the condition repeats on every retry.

**Note:** the delete path itself is correctly wired — `DocumentServiceLeaseStoreManagerInMemory` shares a single `ConcurrentDictionary` between the lease manager (writer) and lease container (reader), and `DocumentServiceLeaseManagerInMemory.DeleteAsync` removes from it. The defect is **not** a missing/duplicated store; it is the **null `FeedRange` on the split path** that prevents delete/create from ever being reached.

## Steps to Reproduce

1. Start a Change Feed Processor using an **in-memory lease container**.
2. Read some documents so leases are established.
3. Stop the processor and **save the lease state** (serialized `MemoryStream`).
4. Restart the processor **from the saved state** (leases restored without `FeedRange`).
5. Trigger a **physical partition split** on the monitored container (e.g., raise provisioned throughput / add data to force a split).
6. Observe repeated errors from the split handler; child partitions are never picked up; the parent lease is never removed.
7. Restart the service → leases re-bootstrap and processing resumes (old symptom clears).

> Emulator note: the Cosmos DB emulator cannot reproduce true parent→child split (410/1002 Gone) events with overlapping child ranges, so this must be reproduced against a real account or via a targeted unit test that drives `HandlePartitionGoneAsync` with a null-`FeedRange` lease.

## Expected Behavior

On a split, the processor should:
- resolve the overlapping child ranges,
- create the child leases,
- delete the parent lease,

**regardless of whether the (possibly rehydrated/legacy) lease has a populated `FeedRange`** — deriving the `FeedRangeEpk` from the lease token / PKRange id when `FeedRange` is null.

## Actual Behavior

`NullReferenceException`/`InvalidCastException` thrown from `PartitionSynchronizerCore.HandlePartitionGoneAsync` before children are created or the parent is deleted; error is caught and repeated; range stalls until process restart.

## Proposed Fix (for discussion)

1. **Guard the cast** in `PartitionSynchronizerCore.HandlePartitionGoneAsync`: when `lease.FeedRange` is null, reconstruct a `FeedRangeEpk` from the lease token / PKRange id before calling `TryGetOverlappingRangesAsync`.
2. **Backfill `FeedRange` on rehydration** in `DocumentServiceLeaseStoreManagerInMemory.DeserializeLeaseState` for leases that only carry a PKRange-id token, so restored in-memory state matches freshly created leases.
3. **(Optional hardening)** Fail fast / log clearly if a lease reaches the split handler with an unresolvable range instead of throwing a raw NRE.

## Test Gaps to Close

- No unit/integration coverage for **in-memory lease container + partition split**. Existing `DocumentServiceLeaseContainerInMemoryTests` cover serialization/round-trip only; `PartitionControllerSplitTests` use a mock lease manager and don't exercise the real in-memory delete/create on split.
- Add a regression test: construct an in-memory lease with `FeedRange == null` (mirroring restored state), drive `HandlePartitionGoneAsync`, and assert child leases are created and the parent lease is removed (no exception).

## Related / Secondary Observations (not blocking)

- **In-memory lease init ergonomics:** initialization/serialization via raw `MemoryStream` + JSON is cumbersome and undocumented. A first-class predefined object model + docs for "initialize from an arbitrary saved position" would significantly improve usability.
- `DocumentServiceLeaseManagerInMemory.DeleteAsync` ignores the `TryRemove` return value — harmless for correctness here, but worth a code-quality follow-up.

## Code Pointers

- `Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs` — `HandlePartitionGoneAsync` (the throw site)
- `Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/LeaseManagement/DocumentServiceLeaseCore.cs` — `FeedRange` with `NullValueHandling.Ignore`
- `Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/LeaseManagement/DocumentServiceLeaseStoreManagerInMemory.cs` — `DeserializeLeaseState`
- `Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/LeaseManagement/DocumentServiceLeaseManagerInMemory.cs` — `CreateLeaseIfNotExistAsync` / `DeleteAsync`
- `Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/FeedManagement/PartitionControllerCore.cs` — `HandlePartitionGoneAsync` (catches and re-notifies)

### PR DESCRIPTION
**Pull request by aavasthy**
# Pull Request Template

### What was broken

1. NullReferenceException on split/merge with a missing FeedRange. When a Change Feed Processor lease loses its  FeedRange , a partition split crashed with  NullReferenceException . The parent lease was never cleaned up and no child leases were created, so that range got stuck until process restart.
2. Continuation token lost on offline split. If a split happened while the processor was stopped, the pre-split lease survived in the (in-memory) lease store. On restart, bootstrap could create fresh, continuation-less leases for the new post-split ranges before the stale parent got a chance to hand off its continuation token, so the range restarted from the beginning.

### Root cause

1.  PartitionSynchronizerCore.HandlePartitionGoneAsync  cast  lease.FeedRange  without a null check.  FeedRange  is serialized with  NullValueHandling.Ignore , so leases rehydrated from an in-memory lease container came back null.
2.  PartitionSynchronizerCore.CreateLeasesAsync  only checked whether a range's own id had a lease — not whether any of its parent ranges still had one.
3.  DocumentServiceLeaseManagerInMemory  never backfilled  FeedRange  on rehydrated leases, unlike  DocumentServiceLeaseManagerCosmos  which has done this on  AcquireAsync  since 2019.

### Fix

1.  DocumentServiceLeaseManagerInMemory.AcquireAsync  now backfills  FeedRange  on rehydrated leases via the routing map cache, mirroring  DocumentServiceLeaseManagerCosmos . Fixes the normal feed-read path before any split happens.
2. If  FeedRange  is still null when a split/merge fires,  HandlePartitionGoneAsync  resolves the replacement ranges via a Parents-based lookup. Splits create per-child leases carrying the parent's continuation token; merges throw so the parent lease and its token stay in place.
3.  CreateLeasesAsync  now skips creating a lease for a range if any of its parents ( PartitionKeyRange.Parents ) still has a lease. Mirrors the Java SDK's  PartitionSynchronizerImpl.createLeases . Leaves the stale parent's own split handling responsible for creating children, so the real continuation token is carried forward.

## Type of change

Please delete options that are not relevant.

- [X] Bug fix (non-breaking change which fixes an issue)
- [] New feature (non-breaking change which adds functionality)
- [] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [] This change requires a documentation update

## Closing issues

To automatically close an issue: closes #IssueNumber

## Changelog

- [X] I have added a changelog entry under `### Unreleased` in `changelog.md`
  for the user-facing impact of this change.
- [ ] No changelog entry is required because this PR is one of:
  documentation-only, test-only, CI / build-only, or a pure internal refactor
  with no observable customer impact.

If the second box is checked, briefly justify here:

>

### CODE REVIEW

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs [hunk 1]
```diff
@@ -123,7 +171,13 @@ await overlappingRanges.ForEachAsync(
                 PartitionKeyRange mergedRange = overlappingRanges[0];
                 DefaultTrace.TraceInformation("Lease {0} merged into {1}", leaseToken, mergedRange.Id);
 
-                DocumentServiceLease newLease = await this.leaseManager.CreateLeaseIfNotExistAsync((FeedRangeEpk)partitionBasedLease.FeedRange, lastContinuationToken);
+                // partitionBasedLease.FeedRange can be null here when this lease was resolved through the
+                // null-FeedRange fallback above (e.g. a rehydrated in-memory lease that never had FeedRange
+                // populated). In that case there is nothing to cast - create the lease directly from the
+                // already-resolved merged range instead of dereferencing a null FeedRange.
+                DocumentServiceLease newLease = partitionBasedLease.FeedRange != null
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs]: I think the original design (PR #2013, which authored all of this code) is: on a merge of A + B => C, each gone parent creates a lease over its own slice carrying its own continuation token. C therefore ends up tiled by N leases with N correct tokens. The existing test `HandlePartitionGoneAsync_PKRangeBasedLease_Merge` pins this, it asserts the PartitionKeyRange overload is called `Times.Never` on the merge path. The new fallback does precisely what that test forbids.

Consider the below scenarios, step by step:

- Parent A (null FeedRange) resolves to merged range C and calls CreateLeaseIfNotExistAsync(C, A_token) =>  a lease named C is created, holding A's token.

- Parent B now handles its own Gone. Internally TryCreateDocumentServiceLeaseAsync does a container.TryAdd(...), which fails ("Some other host created lease for {0}") and returns null.

- newLease == `null` =>  nothing gets enqueued — but the branch still returns (newLeases, true). That true means "delete the current lease", so PartitionControllerCore deletes lease B. B's continuation token is now gone.

- Range C is processed from A's continuation only. Whatever B had not yet processed is skipped, or whatever B had already processed is replayed, depending on whose token was further behind.

An ideal outcome would either:

- resolve the parent's own sub-range and keep using the FeedRangeEpk overload; or
- if it genuinely can't be scoped, return `removeCurrentLease: false` (or throw) so the continuation token survives and the situation stays visible.

At minimum, if newLease == null, do not return `true`.

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs [hunk 2]
```diff
@@ -73,15 +80,56 @@ public override async Task CreateMissingLeasesAsync()
...
+            {
+                if (lease is DocumentServiceLeaseCoreEpk)
+                {
+                    // EPK-based leases have no PartitionKeyRangeId to resolve the range from; the lease token
+                    // itself only encodes the range, which is exactly what's missing.
+                    DefaultTrace.TraceError("Lease {0} is gone but has no FeedRange and is not a PKRange-based lease; the range cannot be resolved", leaseToken);
+                    throw new InvalidOperationException($"Lease {leaseToken} is gone but has no FeedRange and its range cannot be resolved.");
+                }
+
+                // This method only ever runs after the service has already confirmed the partition split or
```
Comment (kundadebdatta) [inline Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs]: nit: Reduce code comments.

File: changelog.md [hunk 3]
```diff
@@ -42,6 +42,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
...
+- [6079](https://github.com/Azure/azure-cosmos-dotnet-v3/pulls/6079) ChangeFeedProcessor: Fixes partition split/merge issues with in-memory leases: a `NullReferenceException` when a lease's `FeedRange` was missing after rehydration (now backfilled on acquire, with a routing-map fallback), and a continuation token being silently discarded when a split happened while the host was offline (the SDK now defers creating a post-split range's lease to its parent's own split handling when the parent lease still exists, instead of racing ahead with a fresh, continuation-less lease).
```
Comment (kundadebdatta) [inline changelog.md]: Link is broken. `/pulls/` should be `/pull/`.

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs (hunk 1 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs]: The latest push takes your throw-so-the-token-survives option. The null-FeedRange handling was lifted into the outer HandlePartitionGoneAsync so the decision happens before we pick a private overload:

1. Look up children of the gone parent via a Parents-based reverse lookup on the full range (the old PartitionKeyRangeId is guaranteed gone from the routing map by the time we get here, so a by-id lookup would always fail).
2. If the lookup returns two or more children, it is a split — proceed into the private overload as before and create one lease per child carrying this parent's continuation token.
3. If it returns exactly one child, it is a merge — throw InvalidOperationException immediately. The parent lease is left untouched, its continuation token is preserved, and the message identifies both the scenario and the lease token for telemetry.
4. If it returns zero children, throw as well — the routing map has not caught up yet, so retrying is the correct behavior.

Because the merge case is rejected up front, the private PartitionKeyRange overload's merge branch can safely assume FeedRange is non-null, and the Times.Never invariant from PR #2013 that HandlePartitionGoneAsync_PKRangeBasedLease_Merge pins still holds

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs (hunk 2 above)
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs]: updated

File: changelog.md (hunk 3 above)
Comment (aavasthy) [reply changelog.md]: updated

[the remark this answers is not in the document]
Comment (aavasthy) [reply changelog.md]: updated

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs [hunk 4]
```diff
@@ -123,7 +171,13 @@ await overlappingRanges.ForEachAsync(
...
                 DefaultTrace.TraceInformation("Lease {0} merged into {1}", leaseToken, mergedRange.Id);
 
-                DocumentServiceLease newLease = await this.leaseManager.CreateLeaseIfNotExistAsync((FeedRangeEpk)partitionBasedLease.FeedRange, lastContinuationToken);
+                // partitionBasedLease.FeedRange can be null here when this lease was resolved through the
+                // null-FeedRange fallback above (e.g. a rehydrated in-memory lease that never had FeedRange
+                // populated). In that case there is nothing to cast - create the lease directly from the
+                // already-resolved merged range instead of dereferencing a null FeedRange.
+                DocumentServiceLease newLease = partitionBasedLease.FeedRange != null
+                    ? await this.leaseManager.CreateLeaseIfNotExistAsync((FeedRangeEpk)partitionBasedLease.FeedRange, lastContinuationToken)
+                    : await this.leaseManager.CreateLeaseIfNotExistAsync(mergedRange, lastContinuationToken);
```
[the remark this answers is not in the document]
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs]: That PKRange-overload fallback is gone. The outer HandlePartitionGoneAsync now throws InvalidOperationException when a null-FeedRange PKRange lease resolves to a single child (merge), so we never reach this branch without a real FeedRange — the parent lease and its continuation token stay in place. Covered by HandlePartitionGoneAsync_PKRangeBasedLease_WithNullFeedRange_ResolvesToSingleMergedRange_Throws (asserts CreateLeaseIfNotExistAsync is called Times.Never on this path).

File: Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs [hunk 5]
```diff
@@ -53,6 +53,13 @@ public override async Task CreateMissingLeasesAsync()
                 FeedRangeEpk.FullRange.Range, 
                 NoOpTrace.Singleton, 
                 forceRefresh: false);
+            // A null result means the routing map could not be resolved (e.g. a transient 404 from the
+            // service, or a stale cache entry). Normalize it to empty here, mirroring the Java SDK's
+            // PartitionSynchronizerImpl#enumPartitionKeyRanges, so a caller never dereferences a null list.
+            // CreateLeasesAsync is a no-op for an empty list, so this is a safe, retryable no-op for this
+            // call; the next scheduled bootstrap/refresh cycle will pick up any missing leases once the
+            // routing map becomes resolvable again.
+            ranges ??= Array.Empty<PartitionKeyRange>();
```
[the remark this answers is not in the document]
Comment (aavasthy) [reply Microsoft.Azure.Cosmos/src/ChangeFeedProcessor/Bootstrapping/PartitionSynchronizerCore.cs]: We now throw InvalidOperationException here instead of normalizing null to empty, so BootstrapperCore.InitializeAsync never reaches MarkInitializedAsync on a transient routing-map failure and the next start retries. Covered by CreateMissingLeases_NullOverlappingRanges_ThrowsInvalidOperationException.

**Review by FabianMeiswinkel — APPROVED** [review]
LGTM - Thanks

**Review by kundadebdatta — APPROVED** [review]
LGTM
